tools/test-runner/src/buck_test.rs

```diff
@@ -0,0 +1,253 @@
+            spec.command[0].clone(),
+            spec.command[1..].to_vec(),
+            spec.env,
+            spec.cwd,
+        );
+        let test = Test {
+            command,
+            name: spec.target,
+            labels: spec.labels,
+            contacts: spec.contacts,
+            kind: TestKind::Shell,
+        };
+        return vec![test]; // i.e; "just run it"
+    }
+
+    /// Evaluates the test based solely on its exit code.
+    pub fn evaluate(result: &mut Child) -> bool {
+        result.wait().unwrap().success() // shouldn't block, it already ran
+    }
+}
+
+/// Refer to https://buck.build/files-and-dirs/buckconfig.html#test.external_runner
+///
+/// A JSON spec needs at least these fields, any others will be ignored.
+#[derive(Clone, Debug, Deserialize)]
+pub struct TestSpec {
+    /// The buck target URI of the test rule.
+    pub target: String,
+
+    /// The type of the test.
+    #[serde(rename(deserialize = "type"))]
+    kind: TestKind,
+
+    /// Command line that should be used to run the test.
+    pub command: Vec<String>,
+
+    /// Environment variables to be defined when running the test.
+    pub env: HashMap<String, String>,
+
+    /// Labels that are defined on the test rule.
+    pub labels: HashSet<String>,
+
+    /// Contacts that are defined on the test rule.
+    pub contacts: HashSet<String>,
+
+    /// Working directory the test should be run from.
+    pub cwd: Option<PathBuf>,
+
+    /// Absolute paths to any files required by this test target.
+    required_paths: Option<Vec<PathBuf>>,
+}
+
+/// Reads test specs from a buck test run description at the given path.
+pub fn read<P: AsRef<Path>>(path: P) -> Result<Vec<TestSpec>> {
+    let path = path.as_ref();
+
+    let file = File::open(path)
+        .with_context(|| format!("Failed to read test specs from {}", path.display()))?;
+
+    let specs: Vec<TestSpec> = serde_json::from_reader(BufReader::new(file))
+        .with_context(|| format!("Failed to parse JSON spec from {}", path.display()))?;
+
+    return Ok(specs);
+}
+
+/// Validates a spec into a (possibly empty) set of runnable tests.
+pub fn validate(spec: TestSpec) -> Result<Vec<Test>> {
+    for &exclude in EXCLUDED_LABELS {
+        if spec.labels.contains(exclude) {
+            return Ok(vec![]);
+        }
+    }
+
+    // we'll collect all validation errors we can find
+    let mut error = String::new();
+
+    if spec.command.len() < 1 {
+        let msg = format!("    Error: Empty command line\n");
+        error.push_str(&msg);
+    }
+
+    // if requirements were specified, we better verify they're there
+    for req in spec.required_paths.clone().unwrap_or(vec![]) {
+        if !req.exists() {
+            let msg = format!("    Error: Missing requirement {}\n", req.display());
+            error.push_str(&msg);
+        }
+    }
+
+    if !error.is_empty() {
+        bail!("Invalid context for test target {}\n{}", spec.target, error);
```

Comment:
I like how you collect all the errors instead of just bailing on the first one :)